src/input/plugins/YtdlInputPlugin.cxx

```
@@ -0,0 +1,87 @@
+#include "config.h"
+#include "lib/ytdl/YtdlParser.hxx"
+#include "YtdlInputPlugin.hxx"
+#include "CurlInputPlugin.hxx"
+#include "../AsyncInputStream.hxx"
+#include "../ProxyInputStream.hxx"
+#include "../InputPlugin.hxx"
+#include "config/ConfigGlobal.hxx"
+#include "config/Block.hxx"
+#include "tag/TagBuilder.hxx"
+#include "tag/Tag.hxx"
+#include "event/Call.hxx"
+#include "event/Loop.hxx"
+#include "thread/Cond.hxx"
+#include "util/Alloc.hxx"
+#include "util/Domain.hxx"
+#include "Log.hxx"
+
+#include <cinttypes>
+
+#include <assert.h>
+#include <string.h>
+
+static constexpr Domain ytdl_domain("youtube-dl");
+
+static void
+input_ytdl_init(const ConfigBlock &block)
+{
+	// TODO: a configuration should allow overriding the curl input plugin
+	// this plugin would then accept http/https URIs as well
+}
+
+static void
+input_ytdl_finish() noexcept
+{
+}
+
+struct YtdlUriFlags {
+	bool notags;
+
+	YtdlUriFlags() : notags(false) { }
+};
+
+static std::string ytdl_uri(const char *uri, YtdlUriFlags* flags) {
+	std::string url;
+	if (!strncmp(uri, "youtube-dl://", 13)) {
+		uri += 13;
+		if (!strncmp(uri, "notag/", 6)) {
+			flags->notags = true;
+			uri += 6;
+		}
+		url = uri;
+	}
+
+	return url;
+}
+
+
+static InputStream *input_ytdl_open(const char *uri, Mutex &mutex, Cond &cond)
+{
+	InputStream *res = nullptr;
```

On error, must throw exception instead of returning `nullptr`.